Add fields to get_image_metadata result

SilensAngelusNex opened the pull request:
Add the `description` and `created` fields to `ImageMetaResponse` so that they will be accessible to callers of `get_image_metadata`.

Fixes #3053

## Sanity check:

* ✔️ Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes

* ✔️ Are you doing the PR on the `next` branch?
* ✔️ Have you created/updated the relevant documentation page(s)?

SilensAngelusNex committed:
Make ImageMetadataResponse creation non-pub

SilensAngelusNex committed:
Make ImageMetadataResponse creation non-pub

SilensAngelusNex committed:
Change get_image_metadata error message context

Include the file extension we're using, and make it clear that we're reading
metadata, not necessarily the image itself.

SilensAngelusNex committed:
Change get_image_metadata error message context

Include the file extension we're using, and make it clear that we're reading
metadata, not necessarily the image itself.

SilensAngelusNex committed:
Move format specific logic into new_svg and new_avif

SilensAngelusNex committed:
Move format specific logic into new_svg and new_avif

SilensAngelusNex committed:
Make helpers take parsed exif data instead of raw

This way we won't have to parse multiple times when calling multiple helpers.

SilensAngelusNex committed:
Make helpers take parsed exif data instead of raw

This way we won't have to parse multiple times when calling multiple helpers.

SilensAngelusNex committed:
Add description and created fields to ImageMetaResponse

SilensAngelusNex committed:
Add description and created fields to ImageMetaResponse

SilensAngelusNex commented:
I made the type for `created` be `String`, but I'd really like to make it something more structured, like `exif::DateTime` (which can handle timezones and has sub-second resolution). I'm not really sure what the best way to define would be that so that Tera can use it. Ideally, we'd still be able to compare it with `<` et al, but it seems like Tera doesn't support that so we'd have to provide comparison functions?

LMK if you think there's a better approach here.

Keats commented:
>  I'm not really sure what the best way to define would be so that Tera can use it. Ideally, we'd still be able to compare it with < et al, but it seems like Tera doesn't support that so we'd have to provide comparison functions?

Nope it's not possible, we would need to expose a specific type in the Value enum in Tera for datetimes, I considered it for tera2 but ended up not doing it because it adds lots of complexity to handle all the tz/various libraries. I will add some tests though, eg `my_date is after(d=other_date)`, `my_date is before(d=other_date)`. Not as nice as `<, >` but good enough

Keats commented on components/imageproc/src/helpers.rs:
```diff
@@ -28,22 +29,44 @@ pub fn fix_orientation(img: &DynamicImage, raw_metadata: Option<Vec<u8>>) -> Opt
+    };
+
+    if bytes.len() == 0 {
+        eprintln!("String value for {:?} has empty Ascii vec.", tag);
```
and is an empty value an error?

SilensAngelusNex committed:
Wrap helpers.rs functions in Result

Bubble errors up to the caller so that error messages have the full
context and so that it's clear when the result means "We successfully
checked the metadata, and should make no changes."

SilensAngelusNex committed:
Wrap helpers.rs functions in Result

Bubble errors up to the caller so that error messages have the full
context and so that it's clear when the result means "We successfully
checked the metadata, and should make no changes."

Keats reviewed (commented):
LGTM, just a quick question

SilensAngelusNex committed:
Change helper functions to support missing image metadata

Change the helper.rs functions to make it easier for meta.rs to only log
to stderr in cases where not being able to get the EXIF metadata fields
is actually an issue. For example, it's not a problem if an image has no
description field or if an image is missing EXIF metadata all together,
but it is a problem if the image has metadata but we're unable to parse
it, or if fields have data, but for an unexpected type.

SilensAngelusNex committed:
Change helper functions to support missing image metadata

Change the helper.rs functions to make it easier for meta.rs to only log
to stderr in cases where not being able to get the EXIF metadata fields
is actually an issue. For example, it's not a problem if an image has no
description field or if an image is missing EXIF metadata all together,
but it is a problem if the image has metadata but we're unable to parse
it, or if fields have data, but for an unexpected type.

SilensAngelusNex commented on components/imageproc/src/helpers.rs:
```diff
@@ -23,27 +67,78 @@ pub fn fix_orientation(img: &DynamicImage, raw_metadata: Option<Vec<u8>>) -> Opt
+    get_string_field(metadata, exif::Tag::ImageDescription, exif::In::PRIMARY)
+}
+
+pub fn get_created_datetime(metadata: Option<&Exif>) -> Result<String, ExifError<'_>> {
```
Good point. I think it'd be weird to not have a CreatedDatetime, just like it'd be weird to not have an Orientation, but I think it's fairly normal to be missing a Description. Moved things around so it's easier to omit error logs for fields that it's normal to be missing and so that we don't log multiple times if the metadata is completely missing or can't be parsed.

Keats replied:
> I think it'd be weird to not have a CreatedDatetime,

Some places will strip all metadata of images, I don't know if they even keep a datetime?

SilensAngelusNex replied:
So do you think none of CreatedDatetime Orientation or Description should print an error if they're missing? Or keep the print for Orientation, but skip it for the other two?

Keats replied:
Everything should be optional

SilensAngelusNex committed:
Remove eprintlns for missing metadata

We exepect to encounter image files missing any/all of these metadata
fields; don't log errors in these cases.
